src/mantine-hooks/src/use-resize-observer/use-resize-observer.ts

```diff
@@ -0,0 +1,58 @@
+import { MutableRefObject, useEffect, useMemo, useRef, useState } from 'react';
+
+type ObserverRect = Pick<
+  DOMRectReadOnly,
+  'x' | 'y' | 'top' | 'left' | 'right' | 'bottom' | 'height' | 'width'
+>;
+type UseResizeObserverResult<T extends HTMLElement = any> = [MutableRefObject<T>, ObserverRect];
+
+const defaultState: ObserverRect = {
+  x: 0,
+  y: 0,
+  width: 0,
+  height: 0,
+  top: 0,
+  left: 0,
+  bottom: 0,
+  right: 0,
+};
+
+export function useResizeObserver<T extends HTMLElement = any>(): UseResizeObserverResult {
+  const frameID = useRef(0);
+  const ref = useRef<T>(null);
+
+  const [rect, setRect] = useState<ObserverRect>(defaultState);
+
+  const observer = useMemo(
+    () =>
+      new ResizeObserver((entries) => {
+        const entry = entries[0];
+
+        if (entry) {
+          cancelAnimationFrame(frameID.current);
+
+          frameID.current = requestAnimationFrame(() => {
+            if (ref.current) {
+              setRect(entry.contentRect);
+            }
+          });
+        }
+      }),
+   []);
+
+  useEffect(() => {
+    if (ref.current) {
+      observer.observe(ref.current);
+    }
+
+    return () => {
+      observer.disconnect();
+
+      if (frameID.current) {
+        cancelAnimationFrame(frameID.current);
+      }
+    };
+  }, [ref.current]);
+
+  return [ref, rect];
```

Comment:
It will be easier with just `return [ref, rect] as const;` instead of defining return type (it should automatically handle generic type)